cmd/csi-api-gen/generators/api.go

```diff
@@ -0,0 +1,432 @@
+
+	// first, re-order the inputs by lengths, so that we always process parent packages first
+	sort.Slice(pkgPaths, func(i, j int) bool {
+		return len(pkgPaths[i]) < len(pkgPaths[j])
+	})
+
+	groups := make(map[string]*groupDefinition)
+
+	for _, pkgPath := range pkgPaths {
+		klog.V(5).Infof("Considering input %s", pkgPath)
+
+		pkg := context.Universe[pkgPath]
+		if pkg == nil {
+			// If the input had no Go files, for example.
+			continue
+		}
+
+		if buildAPIGroupDefinitionFromDocComment(pkgPath, pkg, groups) {
+			// found a +csi-proxy-api-gen comment in the package's doc.go file
+			continue
+		}
+
+		if strings.HasPrefix(pkgPath, csiProxyAPIPath) {
+			// part of the canonical API definitions, under client/api
+			buildCanonicalAPIGroupDefinition(pkgPath, pkg, groups)
+			continue
+		}
+
+		// is this package a version of an API group?
+		parts := strings.Split(pkgPath, "/")
+		parentPkgPath := strings.Join(parts[:len(parts)-1], "/")
+		if definition, present := groups[parentPkgPath]; present {
+			if !apiversion.IsValidVersion(pkg.Name) {
+				klog.Fatalf("Unexpected go package %q, should be of the form \"%s/<version>\", where <version> should be a valid API group version identifier",
+					pkg.Name, parentPkgPath)
+			}
+			klog.V(5).Infof("Found version %q for API group %q", pkg.Name, parts[len(parts)-2])
+			definition.addVersion(pkg)
+		}
+	}
+
+	result := make([]*groupDefinition, len(groups))
+	i := 0
+	for _, group := range groups {
+		// sanity check: every group should have at least one version
+		if len(group.versions) == 0 {
+			klog.Fatalf("API group %q doesn't have any version", group.name)
+		}
+		result[i] = group
+		i++
+	}
+
+	return result
+}
+
+// buildAPIGroupDefinitionFromDocComment looks for a +csi-proxy-api-gen comment in the package's
+// doc.go file, and if it finds one build the corresponding API definition.
+func buildAPIGroupDefinitionFromDocComment(pkgPath string, pkg *types.Package, groups map[string]*groupDefinition) bool {
+	commentTags := types.ExtractCommentTags(tagMarker, pkg.Comments)[tagName]
+
+	if len(commentTags) == 0 {
+		return false
+	}
+
+	definition := newGroupDefinition(pkg.Name, pkg.Path)
+
+	for _, commentTag := range commentTags {
+		commentTag = strings.TrimSpace(commentTag)
+		if commentTag == "" {
+			continue
+		}
+
+		parts := strings.Split(commentTag, ":")
+		if len(parts) != 2 {
+			klog.Fatalf("Malformed comment tag for package %q, should be of the form \"<name>:<value>\", found %q",
+				pkgPath, commentTag)
+		}
+		name := strings.TrimSpace(parts[0])
+		value := strings.TrimSpace(parts[1])
+
+		switch name {
+		case "groupName":
+			definition.name = value
+		case "serverBasePkg":
+			definition.serverBasePkg = canonicalizePkgPath(value)
+		case "clientBasePkg":
+			definition.clientBasePkg = canonicalizePkgPath(value)
+		default:
+			klog.Fatalf("Unknown comment tag %q for package %q", name, pkgPath)
+		}
+	}
+
+	klog.V(5).Infof("Found API group %q", definition.name)
```

Comment:
Here we assume we always find one api group? Is that possible that the returned commentTags are all just empty string? Is that possible that name or value is empty string? Is that valid?

Reply:
If there is a `csi-proxy-api-gen` tag in a package's doc.go file, then it must be an API group, regardless of which additional custom options it has (or doesn't have). So yes, past the
```
	if len(commentTags) == 0 {
		return false
	}
```
check on line 160, we know we've found an API group.